minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,635 @@
+        must be provided.
+
+    Example:
+    >>> from zephyr_utils.node_edge import ZNode, ZShape
+    >>> zn1 = ZNode((5, 2), ZShape(m=5))
+    >>> zn1.neighbors()
+    [ZNode(CartesianCoord(x=4, y=1, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=4, y=3, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=6, y=1, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=6, y=3, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=1, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=9, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=3, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=7, y=2, k=None), shape=ZShape(m=5, t=None))]
+    >>> from zephyr_utils.node_edge import ZNode, ZShape
+    >>> zn1 = ZNode((5, 2), ZShape(m=5))
+    >>> zn1.neighbors(nbr_kind="odd")
+    [ZNode(CartesianCoord(x=3, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=7, y=2, k=None), shape=ZShape(m=5, t=None))]
+    """
+
+    def __init__(
+        self,
+        coord: CartesianCoord | ZephyrCoord | tuple[int],
+        shape: ZShape | tuple[int | None] | None = None,
+        convert_to_z: bool | None = None,
+    ) -> None:
+        self.shape = shape
+
+        if convert_to_z is None:
+            self.convert_to_z = len(coord) in (4, 5)
+        else:
+            self.convert_to_z = convert_to_z
+
+        # convert coord to CartesianCoord or ZephyrCoord
+        if not isinstance(coord, (CartesianCoord, ZephyrCoord)):
+            coord = self.get_coord(coord)
+
+        # convert coord to CartesianCoord
+        if isinstance(coord, ZephyrCoord):
+            coord = zephyr_to_cartesian(coord)
+
+        self.ccoord = coord
+
+    @property
+    def shape(self) -> ZShape | None:
+        """Returns the shape of the Zephyr graph ZNode belongs to."""
+        return self._shape
+
+    @shape.setter
+    def shape(self, new_shape) -> None:
+        """Sets a new value for shape"""
+        if new_shape is None:
+            new_shape = ZShape()
+        elif isinstance(new_shape, tuple):
+            new_shape = ZShape(*new_shape)
+        if not isinstance(new_shape, ZShape):
+            raise TypeError(
+                f"Expected shape to be tuple[int | None] or ZShape or None, got {type(new_shape)}"
+            )
+        if hasattr(self, "_ccoord"):
+            if (self._ccoord.k is None) != (new_shape.t is None):
+                raise ValueError(
+                    f"ccoord, shape must be both quotient or non-quotient, got {self._ccoord, new_shape}"
+                )
+        for var, val in {"m": new_shape.m, "t": new_shape.t}.items():
+            if (val is not None) and (not isinstance(val, int)):
+                raise TypeError(f"Expected {var} to be None or 'int', got {type(val)}")
+        self._shape = new_shape
```

Comment:
Empty lines make things more readable here IMO.
```suggestion

        if not isinstance(new_shape, ZShape):
            raise TypeError(
                f"Expected shape to be tuple[int | None] or ZShape or None, got {type(new_shape)}"
            )

        if hasattr(self, "_ccoord"):
            if (self._ccoord.k is None) != (new_shape.t is None):
                raise ValueError(
                    f"ccoord, shape must be both quotient or non-quotient, got {self._ccoord, new_shape}"
                )

        for var, val in {"m": new_shape.m, "t": new_shape.t}.items():
            if (val is not None) and (not isinstance(val, int)):
                raise TypeError(f"Expected {var} to be None or 'int', got {type(val)}")

        self._shape = new_shape
```